packages/vm/src/context.rs

```diff
@@ -302,33 +312,42 @@ mod test {
 
     const GAS_LIMIT: u64 = 5_000_000;
     const DEFAULT_QUERY_GAS_LIMIT: u64 = 300_000;
-
-    fn make_instance() -> Box<WasmerInstance> {
-        let module = compile(&CONTRACT).unwrap();
+    const TESTING_MEMORY_LIMIT: Size = Size::mebi(16);
+
+    fn make_instance() -> (Env<MS, MQ>, Box<WasmerInstance>) {
+        let env = Env::new(GAS_LIMIT);
+
+        let module = compile(&CONTRACT, TESTING_MEMORY_LIMIT).unwrap();
+        let store = module.store();
+        let i32_to_void = FunctionType::new(vec![Type::I32], vec![]);
+        let i32_to_i32 = FunctionType::new(vec![Type::I32], vec![Type::I32]);
+        let i32i32_to_void = FunctionType::new(vec![Type::I32, Type::I32], vec![]);
+        let i32i32_to_i32 = FunctionType::new(vec![Type::I32, Type::I32], vec![Type::I32]);
+        let i32i32i32_to_i32 =
+            FunctionType::new(vec![Type::I32, Type::I32, Type::I32], vec![Type::I32]);
         // we need stubs for all required imports
         let import_obj = imports! {
-            || { setup_context::<MockStorage, MockQuerier>(GAS_LIMIT) },
             "env" => {
-                "db_read" => Func::new(|_a: u32| -> u32 { 0 }),
-                "db_write" => Func::new(|_a: u32, _b: u32| {}),
-                "db_remove" => Func::new(|_a: u32| {}),
-                "db_scan" => Func::new(|_a: u32, _b: u32, _c: i32| -> u32 { 0 }),
-                "db_next" => Func::new(|_a: u32| -> u32 { 0 }),
-                "query_chain" => Func::new(|_a: u32| -> u32 { 0 }),
-                "canonicalize_address" => Func::new(|_a: u32, _b: u32| -> u32 { 0 }),
-                "humanize_address" => Func::new(|_a: u32, _b: u32| -> u32 { 0 }),
-                "debug" => Func::new(|_a: u32| {}),
+                "db_read" => Function::new(store, &i32_to_i32, |_args: &[Val]| { Ok(vec![Val::I32(0)]) }),
```

Comment:
Ah, no more shorthand that auto/macro-derives the types from a closure?

This looks a lot more verbose, but is also more explicit and less magic.

Reply:
Yeah, a lot of additional runtime variables. At https://github.com/wasmerio/wasmer/pull/1911 there is an attempt to streamline signature definitions.